cmd/state/autoupdate.go

```diff
@@ -0,0 +1,164 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/ActiveState/cli/internal/osutils"
+	"github.com/ActiveState/cli/internal/osutils/lockfile"
+	"github.com/gofrs/flock"
+	"github.com/thoas/go-funk"
+
+	"github.com/ActiveState/cli/internal/appinfo"
+	"github.com/ActiveState/cli/internal/condition"
+	"github.com/ActiveState/cli/internal/config"
+	"github.com/ActiveState/cli/internal/constants"
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/exeutils"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/logging"
+	"github.com/ActiveState/cli/internal/output"
+	"github.com/ActiveState/cli/internal/updater"
+)
+
+const CfgKeyLastCheck = "auto_update_lastcheck"
+
+func autoUpdate(args []string, cfg *config.Instance, out output.Outputer) (bool, error) {
+	defer func() {
+		if err := cfg.Set(CfgKeyLastCheck, time.Now()); err != nil {
+			logging.Error("Failed to store last update check: %s", errs.JoinMessage(err))
+		}
+	}()
+	
+	if !shouldRunAutoUpdate(args, cfg) {
+		return false, nil
+	}
+
+	// Check for available update
+	checker := updater.NewDefaultChecker(cfg)
+	up, err := checker.Check()
+	if err != nil {
+		return false, errs.Wrap(err, "Failed to check for update")
+	}
+	if up == nil {
+		logging.Debug("No update found")
+		return false, nil
+	}
+
+	out.Notice(output.Heading(locale.Tl("auto_update_title", "Auto Update")))
+	out.Notice(locale.Tr("auto_update_to_version", constants.Version, up.Version))
+
+	logging.Debug("Auto updating to %s", up.Version)
+
+	// Protect against multiple updates happening simultaneously
+	fileLock := flock.New(filepath.Join(cfg.ConfigPath(), "install.lock"))
+	lockSuccess, err := fileLock.TryLock()
+	if err != nil {
+		return false, errs.Wrap(err, "Could not create file lock required to install update")
+	}
+	if !lockSuccess {
+		logging.Debug("Another update is already in progress")
+		return false, nil
+	}
+	defer fileLock.Unlock()
+
+	targetDir := filepath.Dir(appinfo.StateApp().Exec())
+	err = up.InstallBlocking(targetDir)
+	if err != nil {
+		log := logging.Error
+		innerErr := errs.InnerError(err)
+		if os.IsPermission(innerErr) {
+			return false, locale.WrapInputError(err, "auto_update_permission_err", "", constants.DocumentationURL, errs.JoinMessage(err))
+		}
+		if errors.As(err, new(*lockfile.AlreadyLockedError)) {
+			log("Auto update failed because the update lock file is already in use")
+			return false, nil
+		}
+		return false, errs.Wrap(err, "Failed to install update")
+	}
+
+	out.Notice(locale.Tr("auto_update_relaunch"))
+	out.Notice("") // Ensure output doesn't stick to our messaging
+	code, err := relaunch(args)
+	if err != nil {
+		return true, errs.WrapExitCode(err, code)
+	}
+
+	return true, nil
+}
+
+func shouldRunAutoUpdate(args []string, cfg *config.Instance) bool {
```

Comment:
I like that it is split up like this, with log messages indicating the reason for why it was skipped. :+1:

Reply:
Yeah the conditional was getting a little unwieldy :p